### PR TITLE
First lab

### PR DESCRIPTION
**Pull request by v-sopov**

### CODE REVIEW

File: lab_solution.py [hunk 1]
```diff
@@ -0,0 +1,832 @@
...
+
+# ### Пункт 3
+
+# Теперь по сгенерированным данным отобразим на графике оба класса (синий - "1", красный - "-1")
+
+# In[5]:
+
+plt.plot(x1, y1, 'b.', x2, y2, 'r.', markersize=5, alpha=0.8)
+plt.axis('equal')
+plt.show()
```
Comment (alex-kharlamov) [inline lab_solution.py:111]: Можно использовать %matplotlib inline

File: lab_solution.py [hunk 2]
```diff
@@ -0,0 +1,832 @@
...
+# ### Пункт 2
+
+# In[4]:
+
+def quadratic_loss_optimization(samples, labels):
+    labels = np.asarray(labels)
+    np.reshape(labels, (labels.shape[0], 1))
+    samples_t = samples.transpose();
+    # pseudoinversion in case the matrix here is not invertible 
+    return np.linalg.pinv((samples_t @ samples)) @ samples_t @ labels
```
Comment (alex-kharlamov) [inline lab_solution.py:100]: Кажется, псевдообратную матрицу нужно было сделать ручками, но не суть важно.

File: lab_solution.py [hunk 3]
```diff
@@ -0,0 +1,832 @@
...
+y = [-(shift + coefs[0]*i)/coefs[1] for i in x]
+plt.subplot(121)
+plt.title('Correct')
+plt.plot(x, y, 'k-', x1, y1, 'b.', x2, y2, 'r.', markersize=5, alpha=0.8)
+plt.axis([-4, 8, -8, 10])
+plt.subplot(122)
+plt.title('Predicted')
+plt.hold(True)
+plt.plot(x, y, 'k-')
+plt.scatter(samples_x[:, 0], samples_x[:, 1], c='b', marker='.', s=50, edgecolors='none', alpha=0.7)
```
Comment (alex-kharlamov) [inline lab_solution.py:166]: Edgecolors=None задано в функции по умолчанию, зачем еще раз передавать дефолт?

File: lab_solution.py [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+plt.plot(x, y_quad, 'b', label='quadratic', linewidth=2)
+plt.plot(x, y_svm, 'g', label='hinge loss', linewidth=2)
+plt.plot(x, y_logistic, 'r', label='logistic', linewidth=2)
+plt.plot([-2, 0, 0, 3], [1, 1, 0, 0], 'y', label='empirical risk', linewidth=2)
+plt.legend(loc='upper center')
+plt.hold(False)
+plt.axis([-2, 2, -1, 3])
+plt.grid()
+plt.axes().set_aspect('equal', adjustable='box')
+plt.show()
```
Comment (alex-kharlamov) [inline lab_solution.py:209]: Может быть, сдвинув легенду в левый нижний угол, было бы чуть лучше видно?

File: lab_solution.py [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+# #### Наискорейший спуск:
+
+# In[17]:
+
+def quadratic_loss_gradient(samples, labels, weights):
+    return 2 / samples.shape[0] * (samples @ weights - labels).transpose() @ samples
+
+def steepest_quadratic_descent(samples, labels, start, step=0.01,
+                         threshold=0.00001, points_log = None, steps_max_number=-1):
+    current_location = np.asarray(start, dtype='float64')[np.newaxis].transpose().copy()
```
Comment (alex-kharlamov) [inline lab_solution.py]: Не очень понятно, зачем нужны `copy`, если сильно нужны, можно было передать как параметр в `np.array`, получается 2 раза вызов одинакового метода.

File: lab_solution.py [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+def get_subsets(sample_set, label_set, subset_size=1):
+    perm = np.random.permutation(len(sample_set))
+    shuffled_samples = sample_set[perm]
+    shuffled_labels = label_set[perm]
+    subsets = []
+    int_bound = len(shuffled_samples) // subset_size
+    for i in range(int_bound):
+        subsets.append((shuffled_samples[i*subset_size:(i+1)*subset_size],
+                        shuffled_labels[i*subset_size:(i+1)*subset_size]))
+    return subsets
```
Comment (alex-kharlamov) [inline lab_solution.py]: 2 раза хранить одни и те же данные? Кажется, генераторы на лету были бы менее беспощадны к оперативке.

File: lab_solution.py [hunk 7]
```diff
@@ -0,0 +1,832 @@
...
+
+
+# ### Получаем данные
+
+# In[20]:
+
+data = genfromtxt('train.csv', delimiter=',')[1:]
```
Comment (alex-kharlamov) [inline lab_solution.py]: Есть сильно более быстрая и удобная альтернатива в виде `pandas.read_csv`

File: lab_solution.py [hunk 8]
```diff
@@ -0,0 +1,832 @@
...
+    return result.dot(samples)
+
+loss_coefs = [x/10 for x in range(1, 10)]
+losses = [[] for i in range(9)]
+sizes = [5, 10, 25, 50, 100, 200, 300, 400, 500, 600, 700, 800]
+simple_size_steps = []
+sgd_log = []
+for size in sizes:
+    current_weights = np.zeros(prepared_samples.shape[1])
+    if size == 300:
```
Comment (alex-kharlamov) [inline lab_solution.py:500]: Что это, зачем это? 300? Почему 300? А другие размеры?

**Review by alex-kharlamov — CHANGES_REQUESTED** [review]
First review before merging.

File: lab_solution.py (hunk 2 above)
Comment (v-sopov) [reply lab_solution.py:100]: Ну, это что-то из категории самостоятельного написания матричного умножения и т.д. - numpy есть, вот нам и счастье

File: lab_solution.py (hunk 1 above)
Comment (v-sopov) [reply lab_solution.py:111]: Какова выгода?

File: lab_solution.py (hunk 3 above)
Comment (v-sopov) [reply lab_solution.py:166]: У функции plot задано, у функции scatter - нет, по умолчанию рисует границы

File: lab_solution.py (hunk 4 above)
Comment (v-sopov) [reply lab_solution.py:209]: В левый нижний - нет. Но в правый верхний, пожалуй, да. Сделаю

File: lab_solution.py (hunk 5 above)
Comment (v-sopov) [reply lab_solution.py]: Подправлю

File: lab_solution.py (hunk 6 above)
Comment (v-sopov) [reply lab_solution.py]: Будут тебе генераторы :)

File: lab_solution.py (hunk 7 above)
Comment (v-sopov) [reply lab_solution.py]: Гляну

File: lab_solution.py (hunk 1 above)
Comment (alex-kharlamov) [reply lab_solution.py:111]: Не нужно каждый раз писать `plt.show`

File: lab_solution.py (hunk 3 above)
Comment (alex-kharlamov) [reply lab_solution.py:166]: image
image
Все-таки задано.

File: lab_solution.py (hunk 8 above)
Comment (v-sopov) [reply lab_solution.py:500]: Мда, сильно извиняюсь за <нехорошее слово>код. Вообще это далее используется для подсчёта экспоненциального сглаживания.
300 выбрано просто как значение "ни туда, ни сюда". Для каждого размера батча тестировать это смысла нет, потому что тренд, как было мною проверено, одинаковый для всех размеров батчей, так что смысла явно тестировать на них особо нет.

File: lab_solution.py (hunk 1 above)
Comment (v-sopov) [reply lab_solution.py:111]: Ну такое...
Предпочитаю иметь более прямой контроль над выводом.

File: lab_solution.py (hunk 3 above)
Comment (v-sopov) [reply lab_solution.py:166]: None != 'none'
None - значение по умолчанию (patch.edgecolor)
'none' - отсутствие границы

Comment (alex-kharlamov) [reply lab_solution.py:166]: хм, постиг, тогда извиняюсь.

File: lab_solution.py (hunk 2 above)
Comment (dasimagin) [reply lab_solution.py:100]: Думаю, что достаточно `np.linalg.pinv((samples)) @ labels`, а теперь надо понятно, почему разделение было правильное ;)

File: lab_solution.py [hunk 9]
```diff
@@ -0,0 +1,831 @@
...
+plt.plot(x1, y1, 'b.', x2, y2, 'r.', markersize=5, alpha=0.8)
+plt.axis('equal')
+plt.show()
+
+
+# ### Пункт 4
+
+# In[6]:
+
+x1_t = x1.reshape((x1.shape[0], 1))
```
Comment (dasimagin) [inline lab_solution.py:118]: Перемудрил ты тут немного, зачем-то в начале расщепил на два вектора, потом столько возни.

File: lab_solution.py [hunk 10]
```diff
@@ -0,0 +1,831 @@
...
+coefs = analitic_result[:-1, 0]
+shift = analitic_result[-1, 0]
+print(coefs, shift)
+
+
+# Простая функция определения класса
+
+# In[7]:
+
+def classify(x, coefs, shift):
```
Comment (dasimagin) [inline lab_solution.py:139]: Там все это прекрасно пишется через матричные выражения ;)

File: lab_solution.py [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+samples_y = samples[labeled_samples == -1]
+
+misses = samples[labeled_samples != labels[:, 0]].shape[0]
+
+
+# Теперь нарисуем на графике прямую, разбивающую точки на классы, и сами классы
+
+# In[8]:
+
+x = [x for x in range(-4, 9)]
```
Comment (dasimagin) [inline lab_solution.py:156]: Для фиксации прямой достаточно двух точек.

File: lab_solution.py [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+    while step_number != steps_max_number:
+        if points_log != None:
+            points_log.append(current_location.copy())
+        gradient = gradient_function(*(current_location.tolist()))
+        gradient *= step
+        current_location -= gradient
+        if math.sqrt(np.dot(gradient, gradient)) < threshold:
+            break
+        step_number += 1
+    if points_log != None:
```
Comment (dasimagin) [inline lab_solution.py:235]: можно просто писать `if points_log:`

File: lab_solution.py [hunk 13]
```diff
@@ -0,0 +1,831 @@
...
+        points_log.append(current_location)
+    return current_location
```
Comment (dasimagin) [inline lab_solution.py:237]: логирование лучше так же через return передавать, в питоне принято возвращать tuple-ы.

File: lab_solution.py [hunk 14]
```diff
@@ -0,0 +1,831 @@
...
+
+
+# Ответ, полученный методом ГС, в данном случае очень близок к аналитическому решению
+
+# ### Метод наискорейшего спуска
+# #### Одномерная оптимизация квадратичной функции потерь по направлению:
+
+# In[17]:
+
+def quadratic_descent_1d(samples, labels, start, direction, step=0.01):
```
Comment (dasimagin) [inline lab_solution.py:383]: Название так себе, да и лучше бы learning_rate возвращать

File: lab_solution.py [hunk 15]
```diff
@@ -0,0 +1,831 @@
...
+
+# In[20]:
+
+def get_subsets(sample_set, label_set, subset_size=1):
+    perm = np.random.permutation(len(sample_set))
+    shuffled_samples = sample_set[perm]
+    shuffled_labels = label_set[perm]
+    int_bound = len(shuffled_samples) // subset_size
+    for i in range(int_bound):
+        yield ((shuffled_samples[i*subset_size:(i+1)*subset_size],
```
Comment (dasimagin) [inline lab_solution.py:452]: зачем столько скобок? кажется, хвост ты теряешь..

File: lab_solution.py [hunk 16]
```diff
@@ -0,0 +1,831 @@
...
+data = data[np.logical_or(data[:, 0] == 0, data[:, 0] == 1)]
+samples = data[:8000, 1:]
+prepared_samples = np.hstack((samples, np.ones((samples.shape[0], 1))))
+labels = data[:8000, 0]
+labels = (labels - 0.5) * 2
+
+test_samples = data[8000:, 1:]
+prepared_test_samples = np.hstack((test_samples, np.ones((test_samples.shape[0], 1))))
+test_labels = data[8000:, 0]
+test_labels = (test_labels - 0.5) * 2
```
Comment (dasimagin) [inline lab_solution.py:472]: Опасна штука. Там может не чистые +1/-1 получится.

File: lab_solution.py (hunk 8 above)
Comment (dasimagin) [reply lab_solution.py:500]: Огромный нечитаемый блок кода :(

File: lab_solution.py [hunk 17]
```diff
@@ -0,0 +1,831 @@
...
+    return result / samples.shape[0]
+
+def logistic_gradient(samples, labels, weights):
+    result = np.exp(np.multiply(-labels, samples.dot(weights)))
+    result = np.divide(result, result + 1)
+    result = np.multiply(result, -labels)
+    result *= 1 / samples.shape[0]
+    return result.dot(samples)
+
+loss_coefs = [x/10 for x in range(1, 10)]
```
Comment (dasimagin) [inline lab_solution.py:493]: для этого есть функция в numpy

File: lab_solution.py [hunk 18]
```diff
@@ -0,0 +1,831 @@
...
+losses = [[] for i in range(9)]
```
Comment (dasimagin) [inline lab_solution.py:494]: `losses = [[]]*9` а еще лучш все через константы провязывать

File: lab_solution.py [hunk 19]
```diff
@@ -0,0 +1,831 @@
...
+sizes = [5, 10, 25, 50, 100, 200, 300, 400, 500, 600, 700, 800]
```
Comment (dasimagin) [inline lab_solution.py:495]: размер чего?

File: lab_solution.py [hunk 20]
```diff
@@ -0,0 +1,831 @@
...
+simple_size_steps = []
+sgd_log = []
+for size in sizes:
+    current_weights = np.zeros(prepared_samples.shape[1])
+    if size == 300:
+        sgd_log.append(current_weights)
+        for loss_coef, loss in zip(loss_coefs, losses):
+            loss.append(logistic_loss(prepared_samples, labels, current_weights))
```
Comment (dasimagin) [inline lab_solution.py:503]: везде ж одинаковые значения, зачем каждый раз считать?

File: lab_solution.py [hunk 21]
```diff
@@ -0,0 +1,831 @@
...
+    steps_number = 0
+    while True:  
+        old_weights = current_weights.copy()
+        for subset in get_subsets(prepared_samples, labels, size):
+            current_weights = gradient_descent(subset[0], subset[1], current_weights,
```
Comment (dasimagin) [inline lab_solution.py:508]: там слишко большая логика внутри накручена, надо попроще функцию использовать

Comment (dasimagin) [reply lab_solution.py:508]: лучше писать так

```
result = function(
    arg1 = val1,
    arg2 = val2,
    ...
)
```

File: lab_solution.py [hunk 22]
```diff
@@ -0,0 +1,831 @@
...
+                                               logistic_gradient, step=1e-4, threshold=1e-4, steps_max_number=1)
+            if size == 300:
+                sgd_log.append(current_weights)
+                for loss_coef, loss in zip(loss_coefs, losses):
+                    loss.append(loss_coef * loss[-1] + (1 - loss_coef) * logistic_loss(subset[0], subset[1], current_weights))
```
Comment (dasimagin) [inline lab_solution.py:513]: И здесь такое же замечание.

File: lab_solution.py [hunk 23]
```diff
@@ -0,0 +1,831 @@
...
+        steps_number += 1
+        if math.sqrt(np.dot(old_weights - current_weights, old_weights - current_weights)) < 1e-4:
```
Comment (dasimagin) [inline lab_solution.py:515]: я думаю в numpy есть евклидова метрика.

File: lab_solution.py (hunk 16 above)
Comment (v-sopov) [reply lab_solution.py:472]: Да вроде не должно. Дробная часть у нас тут является отрицательной степенью двойки, так что стандартные флоаты с этим без округления работают
